Key refactor

tbg opened the pull request:
some more (but less involved) refactoring: mostly renames in what was formerly `RequestHeader`, and the introduction of `keys.RKey` to distinguish keys which have had `KeyAddress` called on before from those which don't.
fixes #2198

tbg committed:
[auto] rename {BatchRequest_->}Header

no manual intervention

tbg committed:
remove (*Span).ContainsKey()

tbg committed:
[auto] rename RequestHeader to Span

no manual intervention.

tbg committed:
[auto] rename in roachpb.Span: EndKey->End, Key->Start

no manual intervention.

tbg committed:
replace keys.Span w/ roachpb.Span

tbg committed:
introduce keys.RKey

for use in key ranges which have had
their local addressing taken care of.

tbg committed:
[auto] rename Span.{Start->Key,End->EndKey}

undoing an earlier change. The rebase is unfortunately
messy, so simply undoing this here.

no manual intervention.

tamird commented on roachpb/data.go:
```diff
@@ -131,6 +132,8 @@ func (k EncodedKey) PrefixEnd() EncodedKey {
 }
 
 // Less compares two keys.
+// TODO(tschottdorf): see whether we actually needs this. We shouldn't be
```
this seems important enough to explore in this PR

tschottdorf replied:
I'll take a look.

tschottdorf replied:
After some rabbit-holing there are 20 call sites left (down from 34). I think they're all redundant but they'll incur enough refactoring to warrant a new PR, so that's what I'd rather do.

tamird commented on storage/store.go:
```diff
@@ -1156,6 +1154,7 @@ func (s *Store) Send(ctx context.Context, ba roachpb.BatchRequest) (*roachpb.Bat
 	for _, union := range ba.Requests {
 		arg := union.GetInner()
 		header := arg.Header()
+		// TODO(tschottdorf): something fishy here
```
more words

tschottdorf replied:
removed (not sure what I thought was fishy - some of these checks could operate on `keys.Range(ba)` instead, but that's a) performance b) gives worse errors).

tbg committed:
"remove" roachpb.KeySlice

it was only required in one test, so that
is now where it lives.

bdarnell commented:
Does this actually fix #2198 or is it just refactoring in preparation for a fix? I didn't notice any tests for the kinds of problems mentioned in #2198 (mixed local/global keys, etc)

tschottdorf commented:
It's essentially a code-audit (btw, I'm going to add another commit threading `RKey` through more). I opened that issue because I realized that we were being more than sloppy with local adressing; it should be fairly solid after I'm done with this PR: all user keys are unresolved, and for the most part we'll be dealing with resolved keys internally (to look up ranges, etc). One thing I want to write tests for (and currently there are certainly bugs with this) is to make sure that range-local and range-global ranges must not be mixed within a batch (`\localX`-`A` could resolve to `[X,A)` which is empty, wreaking havoc, etc), and of course that both endpoints of a key range are of the same type.

tbg committed:
thread RKey through more of the code

tbg committed:
rename {keys->roachpb}.RKey

tbg committed:
rename keys.{KeyAddress->Addr}

tbg committed:
RKey for Range{Tree*,Descriptor}, Txn.Key, key consts

tbg committed:
disallow requests crossing local->global boundary

such as Scan([KeyMin, "a"]) or Scan("\lclX", "a")

tbg committed:
test keys.Range(roachpb.BatchRequest)

tbg committed:
another test for local->global boundary crossing

tschottdorf commented:
ok, I think I've added all the code that I want to add, PTAL.

Removed some TODOs by adding tests (in particular for/against requests crossing from range-local into range-global addressing space), and threaded `RKey` through (most of the) code that requires it.
I'll leave #2198 open but only to potentially remove `(*Key).Less`; there are still a bunch of call sites best refactored separately.

Regarding renaming of `RKey` and consistently using `keys.Addr` (formerly `KeyAddress`) I haven't made any changes yet. A lot of explicit casts in the tests have been dropped (since `RangeDescriptor` now contains `RKey`s) so that I don't really feel the need to use `keys.Addr()` in general. But I'm not strongly opposed. What's your preference @bdarnell?

tbg committed:
continue work based on feedback

- do not cast user key as RKey
- some assertions that
- introduce RKey{Max,Min}, RKey(Max|Min).Key() -> Key(Max|Min)

bdarnell commented:
@mrtracy and I were talking about this change at lunch and we have some serious reservations about it. `keys.Addr` is lossy, and it should only be used to look up range IDs and boundaries. Everything else should still be using raw keys. The breadth of this change suggests that we're using resolved keys in way too many places (does this mean that user keys could collide with local keys that have had LocalRangePrefix stripped?)

I think that in general, methods like `Store.LookupReplica` should take a `roachpb.Key` and be responsible for converting it to an `RKey`. Range descriptors should only contain keys for which the `Key` and `RKey` are the same so it doesn't matter too much which type they contain, although I'd say that at least AdminSplit should take `Keys` instead of `RKey` (just to hide the `RKey` type from the higher-level interfaces). `roachpb.Transaction` and the `RangeTree` messages should also contain `Keys`.

tschottdorf commented:
I'm not convinced about these reservations.

> keys.Addr is lossy, and it should only be used to look up range IDs and boundaries.

Yep, that's the intended scope of the change.

> The breadth of this change suggests that we're using resolved keys in way too many places (does this mean that user keys could collide with local keys that have had LocalRangePrefix stripped?)

To me, the breadth of this change suggests that most of the time we handle keys, we handle **resolved** keys (as in, where does this request go, where do I need to split this range). Maybe my unfortunate choice to introduce a new `RKey` instead of a new `UnresolvedKey` triggered this impression.

> (does this mean that user keys could collide with local keys that have had LocalRangePrefix stripped?)

no, user keys are always `roachpb.Key`, so they don't have anything stripped. If there's anything in the diff suggesting that (@bdarnell already pointed out one instance in `RunLsRanges`), please let me know.

>  I think that in general, methods like Store.LookupReplica should take a roachpb.Key and be responsible for converting it to an RKey

I don't necessarily disagree, it's just that we look up the results of a batch key range, and that's already resolved. And with your suggestion I can call `LookupReplica(\lclA, B)`; it's really a bit more responsibility to handle potentially unresolved keys. The caller usually knows much better what they have, and keys don't need to be resolved often (`Addr` has 16 call sites, but most of those from batch-related handling in `DistSender`).

> Range descriptors should only contain keys for which the Key and RKey are the same so it doesn't matter too much which type they contain

but it does, the one tells you it's not a local key and the other keeps you guessing. It's a bit like passing a value whose type you know as an `interface{}`. What sense does a range descriptor make when it can contain a range-local key? Seems like a welcome win to have the type system help you out, plus we almost exclusively deal with descriptors in the context of addressing - that is, `RKey`s.

> although I'd say that at least AdminSplit should take Keys instead of RKey (just to hide the RKey type from the higher-level interfaces).

but it does? All of the proto messages hold their keys in a `roachpb.Span`, which is just two `roachpb.Key`s.

> roachpb.Transaction and the RangeTree messages should also contain Keys

again, `Txn.Key` really points to a `Range`, and if you put a `roachpb.Key` there you can never quite be sure that some call path just forgot to resolve the addressing. I can't see the upside of that suggestion. Same for RangeTree.

I must be completely missing the point of the criticism. If the typing gets too tedious, let's just hop on a call tomorrow.

bdarnell commented:
> most of the time we handle keys, we handle resolved keys

I pulled this branch and count 478 instances of `roachpb.RKey` and 909 instances of `roachpb.Key`. So we're still more likely to handle unresolved keys. When we call `keys.Addr()` we always keep the original unresolved key (or at least we should) because that's what's going to matter by the time we get to the storage layer, and you can't reverse `keys.Addr()`.

> again, Txn.Key really points to a Range, and if you put a roachpb.Key there you can never quite be sure that some call path just forgot to resolve the addressing

This change has actually pointed out a bug in the way we use `Txn.Key`: We store the `RKey` without reference to its original `Key`, so in order to push a transaction we have to convert it back to a key with `RKey.Key` (store.go:1360). This means that we will end up calling `keys.Addr` on a key that has already been resolved. This could give incorrect results if first key of the transaction ends up starting with `RangeLocalPrefix`. Or, if a user key looks like a range descriptor key with the local prefix removed, transactions on that key could conflict with transactions on the range descriptor. The transaction record should instead store the original `Key`, and resolve it with `keys.Addr` when needed.

I suspect that most uses of `RKey.Key` could lead to similar bugs. I think we should get rid of that method and make sure that whenever we need a `Key`, we still have the original key, and only use `RKey` in places where we're OK with the one-way transformation. Am I making more sense now? If not let's talk after the team meeting tomorrow.

tschottdorf commented:
> I pulled this branch and count 478 instances of roachpb.RKey and 909 instances of roachpb.Key. So we're still more likely to handle unresolved keys

That depends on your definition of "handle": grep out generated code, tests and `./storage/engine` (which doesn't even know about `RKey`) and you end up about even (~120). What I mean though is that we rarely do anything with `roachpb.Key`, it's mostly just passed through.

> This means that we will end up calling keys.Addr on a key that has already been resolved.
Agreed that it's semantically wrong and that `Txn.Key` is better off as a `Key`.

> This could give incorrect results if first key of the transaction ends up starting with RangeLocalPrefix

No, you'd run into trouble if the transaction addresses (as first write) a local key wrapped in a local key (`\lcl\lclx`) which would put the txn entry at `\lclx` but the Push would look for range `x`. That borders the byzantine but yes, it needs to be fixed. Fortunately easy in that case.

> I suspect that most uses of RKey.Key could lead to similar bugs. I think we should get rid of that method and make sure that whenever we need a Key, we still have the original key, and only use RKey in places where we're OK with the one-way transformation.

I'm doing an audit now, but I think there will be some instances left in which we want to convert explicitly. Think of RangeDescriptor, for example: I've changed its keys to `RKey` since a `RangeDescriptor` must never ever contain a local key (switching back to `Key` wouldn't do much good except for hiding that problem), and you frequently need to compare those keys - but comparing `roachpb.Key` is actually not really defined (`\lclZ < B`)

There shouldn't ever be a doubly-local key, by the way: local addressing is highly internal and local keys are never "hand-made" (they always go through, e.g., `keys.MakeRangeIDKey`, and that method takes `RKey` arguments). Splits at such keys are also highly undefined (going to make sure they actually can't happen). If we want to safeguard further, we could have `Addr()` return an error if the decoded key isn't idempotent under `Addr()` itself.

tbg committed:
Transaction.Key: {R->}Key

tbg committed:
MetaScanBounds returns Key

tbg committed:
simplify range data iterator key ranges

tbg committed:
introduce AsRawKey() to migrate to from Key()

tbg committed:
clean up kv.truncate() wrt local keys

tbg committed:
change some key constants' type

tbg committed:
remove marshalling of RKey

and change more key types

tbg committed:
remove (Key).Less

tschottdorf commented:
ok, I dropped `(Key).Less` and migrated/inspected all uses of `.Key()` to `ToRawKey()` of which there are 20 left (excluding tests, see https://gist.github.com/tschottdorf/751d3958bc5141a12e8b).

bdarnell commented:
I've been looking at [this link](https://github.com/tschottdorf/cockroach/compare/cdd3782dee5992ec97acdc41f600a74b0e12af79...tschottdorf:key_refactor) to see the entire branch minus the uninteresting renames at the beginning. I think reviewing this commit-by-commit has made it look more invasive than it is.

I think this change is generally an improvement, since it makes things explicit and highlights areas where we were already being sloppy about the resolved or unresolved status of keys. It's probably best to get it merged and we can make any other required changes later.

I still feel like the `RKey` type is being used in places where it probably shouldn't. It's being used for two different meanings: a transformed key, with the local prefix stripped off if present, and a constrained key, which is guaranteed not to have a local prefix. I believe the type system is the wrong way to enforce the latter. For example, many split-related functions take `RKeys` to indicate that keys with a local prefix are not valid split points, but since the canonical way to convert a `Key` to an `RKey` is `keys.Addr` it implies that you can split at arbitrary points as long as you resolve them first.

tbg committed:
unexport key constants

and change some types to Key or []byte as applicable.

tschottdorf commented:
> I still feel like the RKey type is being used in places where it probably shouldn't.

That's true, the last commits have been mostly `RKey`->`Key`. Since we really don't have any uses for doubly-local keys (and since not every local key is handled by `Addr`: `Addr` only takes range-local keys and will explode for anything else) maybe a soothing fix is to have `Addr` return an error if the unwrapped key again has a local portion. That way, an `RKey` is "guaranteed" non-local.

But as you suggested I'd rather not add on this PR more.

bdarnell commented:
LGTM
